CI fails for distributed smoke tests: RuntimeError: aten.mm.default got mixed torch.Tensor and DTensor

CI fails for distributed smoke tests: https://github.com/huggingface/trl/actions/runs/27536646758/job/81387688470
> RuntimeError: aten.mm.default got mixed torch.Tensor and DTensor, need to convert all torch.Tensor to DTensor before calling distributed operators!

```python
FAILED tests/distributed/test_distributed.py::TestDistributed::test_sft_peft[fsdp2] - AssertionError: assert 1 == 0
 +  where 1 = CompletedProcess(args=['accelerate', 'launch', '--config_file', PosixPath('/tmp/pytest-of-root/pytest-0/test_sft_peft_fsdp2_0/data/accelerate_configs/fsdp2.yaml'), 'trl/scripts/sft.py', '--output_dir', '/tmp/pytest-of-root/pytest-0/test_sft_peft_fsdp2_0', '--model_name_or_path', 'trl-internal-testing/tiny-Qwen2ForCausalLM-2.5', '--dataset_name', 'trl-internal-testing/zen', '--dataset_config', 'standard_language_modeling', '--use_peft'], returncode=1).returncode
```

Stacktrace:
```python
>       run_command(
            [
                "accelerate", "launch", "--config_file", get_config_path(config), "trl/scripts/sft.py",
                "--output_dir", self.tmp_dir,
                "--model_name_or_path", "trl-internal-testing/tiny-Qwen2ForCausalLM-2.5",
                "--dataset_name", "trl-internal-testing/zen",
                "--dataset_config", "standard_language_modeling",
                "--use_peft",
            ],
            os.environ.copy(),
        )

tests/distributed/test_distributed.py:213: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

command = ['accelerate', 'launch', '--config_file', PosixPath('/tmp/pytest-of-root/pytest-0/test_sft_peft_fsdp2_0/data/accelerate_configs/fsdp2.yaml'), 'trl/scripts/sft.py', '--output_dir', ...]
env = {'NV_LIBCUBLAS_VERSION': '12.8.4.1-1', 'NVIDIA_VISIBLE_DEVICES': 'GPU-1875d08e-72bf-a64c-3d5f-56f12b63f8b2,GPU-725befcf-73aa-5eed-34f1-112b8d00c11a', 'KUBERNETES_SERVICE_PORT_HTTPS': '443', 'NV_NVML_DEV_VERSION': '12.8.90-1', ...}

    def run_command(command: list[str], env: dict[str, str]) -> None:
        result = subprocess.run(command, env=env, cwd=ROOT)
>       assert result.returncode == 0
E       AssertionError: assert 1 == 0
E        +  where 1 = CompletedProcess(args=['accelerate', 'launch', '--config_file', PosixPath('/tmp/pytest-of-root/pytest-0/test_sft_peft_fsdp2_0/data/accelerate_configs/fsdp2.yaml'), 'trl/scripts/sft.py', '--output_dir', '/tmp/pytest-of-root/pytest-0/test_sft_peft_fsdp2_0', '--model_name_or_path', 'trl-internal-testing/tiny-Qwen2ForCausalLM-2.5', '--dataset_name', 'trl-internal-testing/zen', '--dataset_config', 'standard_language_modeling', '--use_peft'], returncode=1).returncode

tests/distributed/test_distributed.py:32: AssertionError
----------------------------- Captured stderr call -----------------------------
[RANK 0] `loss_type='chunked_nll'` under FSDP2 with `reshard_after_forward=True` is significantly slower than necessary due to per-chunk all-gathers of `lm_head.weight`. Consider passing `--fsdp_reshard_after_forward false` to `accelerate launch` (or equivalent in your FSDP config).
[transformers] The tokenizer has new PAD/BOS/EOS tokens that differ from the model config and generation config. The model config and generation config were aligned accordingly, being updated with the tokenizer's values. Updated tokens: {'eos_token_id': 151645, 'bos_token_id': None, 'pad_token_id': 151643}.
[transformers] The tokenizer has new PAD/BOS/EOS tokens that differ from the model config and generation config. The model config and generation config were aligned accordingly, being updated with the tokenizer's values. Updated tokens: {'eos_token_id': 151645, 'bos_token_id': None, 'pad_token_id': 151643}.
[rank1]: Traceback (most recent call last):
[rank1]:   File "/__w/trl/trl/trl/scripts/sft.py", line 140, in <module>
[rank1]:     main(script_args, training_args, model_args, dataset_args)
[rank1]:   File "/__w/trl/trl/trl/scripts/sft.py", line 112, in main
[rank1]:     trainer.train()
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/transformers/trainer.py", line 1433, in train
[rank1]:     return inner_training_loop(
[rank1]:            ^^^^^^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/transformers/trainer.py", line 1515, in _inner_training_loop
[rank1]:     self._run_epoch(
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/transformers/trainer.py", line 1743, in _run_epoch
[rank1]:     tr_loss_step = self.training_step(model, inputs, num_items_in_batch)
[rank1]:                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/trl/trainer/sft_trainer.py", line 1741, in training_step
[rank1]:     return super().training_step(*args, **kwargs)
[rank1]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/transformers/trainer.py", line 1915, in training_step
[rank1]:     loss = self.compute_loss(model, inputs, num_items_in_batch=num_items_in_batch)
[rank1]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/trl/trainer/sft_trainer.py", line 1639, in compute_loss
[rank1]:     (loss, outputs) = super().compute_loss(
[rank1]:                       ^^^^^^^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/transformers/trainer.py", line 1987, in compute_loss
[rank1]:     outputs = model(**inputs)
[rank1]:               ^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1778, in _wrapped_call_impl
[rank1]:     return self._call_impl(*args, **kwargs)
[rank1]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1884, in _call_impl
[rank1]:     return inner()
[rank1]:            ^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1832, in inner
[rank1]:     result = forward_call(*args, **kwargs)
[rank1]:              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/peft/peft_model.py", line 1993, in forward
[rank1]:     return self.base_model(
[rank1]:            ^^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1778, in _wrapped_call_impl
[rank1]:     return self._call_impl(*args, **kwargs)
[rank1]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1789, in _call_impl
[rank1]:     return forward_call(*args, **kwargs)
[rank1]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/peft/tuners/tuners_utils.py", line 330, in forward
[rank1]:     return self.model.forward(*args, **kwargs)
[rank1]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/trl/trainer/sft_trainer.py", line 312, in _chunked_ce_forward
[rank1]:     loss, num_correct_tokens, entropy_sum, num_valid_tokens = _chunked_cross_entropy_loss(
[rank1]:                                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/trl/trainer/sft_trainer.py", line 217, in _chunked_cross_entropy_loss
[rank1]:     chunk_loss, chunk_correct, chunk_entropy = torch.utils.checkpoint.checkpoint(
[rank1]:                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/_compile.py", line 54, in inner
[rank1]:     return disable_fn(*args, **kwargs)
[rank1]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/_dynamo/eval_frame.py", line 1298, in _fn
[rank1]:     return fn(*args, **kwargs)
[rank1]:            ^^^^^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/utils/checkpoint.py", line 518, in checkpoint
[rank1]:     ret = function(*args, **kwargs)
[rank1]:           ^^^^^^^^^^^^^^^^^^^^^^^^^
[rank1]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/trl/trainer/sft_trainer.py", line 114, in _chunk
[rank1]:     logits = h.float() @ w.float().t()
[rank1]:              ~~~~~~~~~~^~~~~~~~~~~~~~~
[rank1]: RuntimeError: aten.mm.default got mixed torch.Tensor and DTensor, need to convert all torch.Tensor to DTensor before calling distributed operators!
[rank0]: Traceback (most recent call last):
[rank0]:   File "/__w/trl/trl/trl/scripts/sft.py", line 140, in <module>
[rank0]:     main(script_args, training_args, model_args, dataset_args)
[rank0]:   File "/__w/trl/trl/trl/scripts/sft.py", line 112, in main
[rank0]:     trainer.train()
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/transformers/trainer.py", line 1433, in train
[rank0]:     return inner_training_loop(
[rank0]:            ^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/transformers/trainer.py", line 1515, in _inner_training_loop
[rank0]:     self._run_epoch(
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/transformers/trainer.py", line 1743, in _run_epoch
[rank0]:     tr_loss_step = self.training_step(model, inputs, num_items_in_batch)
[rank0]:                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/trl/trainer/sft_trainer.py", line 1741, in training_step
[rank0]:     return super().training_step(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/transformers/trainer.py", line 1915, in training_step
[rank0]:     loss = self.compute_loss(model, inputs, num_items_in_batch=num_items_in_batch)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/trl/trainer/sft_trainer.py", line 1639, in compute_loss
[rank0]:     (loss, outputs) = super().compute_loss(
[rank0]:                       ^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/transformers/trainer.py", line 1987, in compute_loss
[rank0]:     outputs = model(**inputs)
[rank0]:               ^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1778, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1884, in _call_impl
[rank0]:     return inner()
[rank0]:            ^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1832, in inner
[rank0]:     result = forward_call(*args, **kwargs)
[rank0]:              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/peft/peft_model.py", line 1993, in forward
[rank0]:     return self.base_model(
[rank0]:            ^^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1778, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1789, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/peft/tuners/tuners_utils.py", line 330, in forward
[rank0]:     return self.model.forward(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/trl/trainer/sft_trainer.py", line 312, in _chunked_ce_forward
[rank0]:     loss, num_correct_tokens, entropy_sum, num_valid_tokens = _chunked_cross_entropy_loss(
[rank0]:                                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/trl/trainer/sft_trainer.py", line 217, in _chunked_cross_entropy_loss
[rank0]:     chunk_loss, chunk_correct, chunk_entropy = torch.utils.checkpoint.checkpoint(
[rank0]:                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/_compile.py", line 54, in inner
[rank0]:     return disable_fn(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/_dynamo/eval_frame.py", line 1298, in _fn
[rank0]:     return fn(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/utils/checkpoint.py", line 518, in checkpoint
[rank0]:     ret = function(*args, **kwargs)
[rank0]:           ^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/__w/trl/trl/.venv/lib/python3.12/site-packages/trl/trainer/sft_trainer.py", line 114, in _chunk
[rank0]:     logits = h.float() @ w.float().t()
[rank0]:              ~~~~~~~~~~^~~~~~~~~~~~~~~
[rank0]: RuntimeError: aten.mm.default got mixed torch.Tensor and DTensor, need to convert all torch.Tensor to DTensor before calling distributed operators!
E0615 09:34:31.953000 6676 torch/distributed/elastic/multiprocessing/api.py:988] failed (exitcode: 1) local_rank: 0 (pid: 6809) of binary: /__w/trl/trl/.venv/bin/python3
Traceback (most recent call last):
  File "/__w/trl/trl/.venv/bin/accelerate", line 10, in <module>
    sys.exit(main())
             ^^^^^^
  File "/__w/trl/trl/.venv/lib/python3.12/site-packages/accelerate/commands/accelerate_cli.py", line 50, in main
    args.func(args)
  File "/__w/trl/trl/.venv/lib/python3.12/site-packages/accelerate/commands/launch.py", line 1394, in launch_command
    multi_gpu_launcher(args)
  File "/__w/trl/trl/.venv/lib/python3.12/site-packages/accelerate/commands/launch.py", line 1023, in multi_gpu_launcher
    distrib_run.run(args)
  File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/distributed/run.py", line 1007, in run
    elastic_launch(
  File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/distributed/launcher/api.py", line 191, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/distributed/launcher/api.py", line 371, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError: 
============================================================
trl/scripts/sft.py FAILED
------------------------------------------------------------
Failures:
[1]:
  time      : 2026-06-15_09:34:31
  host      : aws-g5-12xlarge-cache-use1-public-80-phcjz-runner-bj7c-workflow
  rank      : 1 (local_rank: 1)
  exitcode  : 1 (pid: 6810) 
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2026-06-15_09:34:31
  host      : aws-g5-12xlarge-cache-use1-public-80-phcjz-runner-bj7c-workflow
  rank      : 0 (local_rank: 0)
  exitcode  : 1 (pid: 6809) 
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
```

--- a/trl/trainer/sft_trainer.py
+++ b/trl/trainer/sft_trainer.py
@@ -237,7 +237,9 @@ def _chunked_cross_entropy_loss(
     return loss, correct, entropy_sum, n_valid_tensor
 
 
-def _patch_chunked_ce_lm_head(model: torch.nn.Module, chunk_size: int, is_vlm: bool = False) -> None:
+def _patch_chunked_ce_lm_head(
+    model: torch.nn.Module, chunk_size: int, is_vlm: bool = False, with_peft: bool = False
+) -> None:
     """
     Patch `model.forward` to compute the LM loss via [`_chunked_cross_entropy_loss`].
 
@@ -309,16 +311,26 @@ def _patch_chunked_ce_lm_head(model: torch.nn.Module, chunk_size: int, is_vlm: b
         )
         hidden_states = outputs.last_hidden_state
 
+        lm_head_weight = lm_head.weight
+        lm_head_bias = lm_head.bias
+        # Under FSDP2 + PEFT, the outer FSDP module wraps PeftModel while lm_head.weight
+        # belongs to the nested causal-LM FSDP module, which is never entered directly inside
+        # _chunked_ce_forward. The weight therefore stays in Shard(0) placement; full_tensor()
+        # all-gathers the shards into the complete weight before the chunked matmul.
+        if with_peft and isinstance(lm_head_weight, torch.distributed.tensor.DTensor):
+            lm_head_weight = lm_head_weight.full_tensor()
+            if lm_head_bias is not None:
+                lm_head_bias = lm_head_bias.full_tensor()
         loss, num_correct_tokens, entropy_sum, num_valid_tokens = _chunked_cross_entropy_loss(
             hidden_states,
-            lm_head.weight,
+            lm_head_weight,
             chunk_size,
             labels=labels,
             shift_labels=shift_labels,
             num_items_in_batch=num_items_in_batch,
             logit_scale=logit_scale,
             final_logit_softcapping=final_logit_softcapping,
-            lm_head_bias=lm_head.bias,
+            lm_head_bias=lm_head_bias,
         )
 
         aux_loss = None
@@ -1307,7 +1319,9 @@ class SFTTrainer(_BaseTrainer):
                             "`lm_head` from `target_modules`, or switch to `loss_type='nll'`. If this is a real use "
                             "case for you, please open an issue at https://github.com/huggingface/trl/issues."
                         )
-                _patch_chunked_ce_lm_head(target, chunk_size=_CHUNKED_LM_HEAD_CHUNK_SIZE, is_vlm=self._is_vlm)
+                _patch_chunked_ce_lm_head(
+                    target, chunk_size=_CHUNKED_LM_HEAD_CHUNK_SIZE, is_vlm=self._is_vlm, with_peft=is_peft_model(model)
+                )
             else:
                 raise ValueError(
                     f"Invalid `loss_type` {args.loss_type} passed. Supported values are 'nll', 'dft', and "

